docs/examples/napari/napari_img_math.md

```diff
@@ -0,0 +1,210 @@
+
+[⬇️](#the-magic-part) Create a `magicgui` widget that can be used in another
+program (napari)
+
+[⬇️](#create-dropdowns-with-enums) Use an `Enum` to create a dropdown menu
+
+[⬇️](#connect-event-listeners-for-interactivity) Connect some event listeners to
+create interactivity.
+
+## code
+
+*Code follows, with explanation below... You can also [get this example at
+github](https://github.com/napari/magicgui/blob/master/examples/napari_image_arithmetic.py).*
+
+```{code-block} python
+---
+lineno-start: 1
+emphasize-lines: 31, 39
+---
+
+from enum import Enum
+
+import numpy
+import napari
+from napari.layers import Image
+
+from magicgui import magicgui
+
+class Operation(Enum):
+    """A set of valid arithmetic operations for image_arithmetic.
+
+    To create nice dropdown menus with magicgui, it's best
+    (but not required) to use Enums.  Here we make an Enum
+    class for all of the image math operations we want to
+    allow.
+    """
+    add = numpy.add
+    subtract = numpy.subtract
+    multiply = numpy.multiply
+    divide = numpy.divide
+
+
+with napari.gui_qt():
+    # create a viewer and add a couple image layers
+    viewer = napari.Viewer()
+    viewer.add_image(numpy.random.rand(20, 20), name="Layer 1")
+    viewer.add_image(numpy.random.rand(20, 20), name="Layer 2")
+
+    # here's the magicgui!  We also use the additional
+    # `call_button` option
+    @magicgui(call_button="execute")
+    def image_arithmetic(
+        layerA: Image, operation: Operation, layerB: Image
+    ) -> Image:
+        """Add, subtracts, multiplies, or divides to image layers."""
+        return operation.value(layerA.data, layerB.data)
+
+    # add our new magicgui widget to the viewer
+    viewer.window.add_dock_widget(image_arithmetic)
+
+    # keep the dropdown menus in the gui in sync with the layer model
+    viewer.layers.events.inserted.connect(image_arithmetic.reset_choices)
+    viewer.layers.events.removed.connect(image_arithmetic.reset_choices)
+```
+
+## walkthrough
+
+We're going to go a little out of order so that the other code makes more sense.
+Let's start with the actual function we'd like to write to do some image
+arithmetic.
+
+### the function
+
+Our function takes two `napari` [Image
+layers](https://napari.org/tutorials/fundamentals/image), and some mathematical
+operation (we'll restrict the options using an {class}`~enum.Enum`).  When called, our
+function calls the selected operation on the two layers (which each have a
+`data` attribute that stores the array info).
+
+```python
+def image_arithmetic(layerA, operation, layerB):
+    return operation.value(layerA.data, layerB.data)
+```
+
+#### type annotations
+
+`magicgui` works particularly well with [type
+annotations](https://docs.python.org/3/library/typing.html), and allows
+third-party libraries to register widgets for handling their custom types.
+`napari` provides support for `magicgui` by registering a dropdown menu whenever
```

Comment:
perhaps link to the relevant part of the napari code base for this? or to the `register_type` function, at least.

The other question I had while reading this tutorial is whether I can set the name of the output layer to anything other than result, and/or whether I can prevent it from being overwritten in-place (rather, each execution creates a new layer).

Reply:
yes you can name the output layer, by providing a `{"name": ...}` key in your layer data tuple...
However, I think creating a new layer on each call would require some improvements to `napari.utils._magicgui` (which needs some attention anyway in https://github.com/napari/napari/issues/2028)